### PR TITLE
(#61) Extends support for GetApplicationName()

### PR DESCRIPTION
**Pull request by Originalutter**
Tests needs to be confirmed or updated.

C:\\Projects\\Temp\\My.Console.App\\bin\\Debug\\RawException.vshost.exe => My.Console.App
C:\\Projects\\Temp\\My.Console.App\\1.2.3.456\\bin\\Debug\\RawException.vshost.exe => My.Console.App
C:\\Projects\\Temp\\My.Application\\My.Application.Web\\bin => My.Application.Web
C:\\Projects\\Temp\\My.Application\\My.Application.Web\\1.2.3.456\\bin => My.Application.Web
C:\\Application\\PRODUCTION\\My.Windows.Service\\ => My.Windows.Service
C:\\Application\\PRODUCTION\\My.Windows.Service\\1.2.3.456\\ => My.Windows.Service

Also, I assumed that the dots should be kept in the application name

### CODE REVIEW

File: src/RawRabbit/Common/NamingConventions.cs [hunk 1]
```diff
@@ -66,13 +67,34 @@ private string GetSubscriberQueueSuffix(Type messageType)
...
-				.Last(d => !_disallowedDirectoryNames.Contains(d))
-				.Split('.')
-				.Last();
+			return GetApplicationNameFromUrl(Directory.GetCurrentDirectory());
+		}
+
+		public string GetApplicationNameFromUrl(string url)
+		{
+			var consoleOrIISHostedAppRegex = new Regex(@".*\\(?<ApplicationName>.*[a-zA-Z][^\\]*)(\\[0-9.]*)?\\bin.*");
+			var match = consoleOrIISHostedAppRegex.Match(url);
```
Comment (pardahlman) [inline src/RawRabbit/Common/NamingConventions.cs]: According to [https://msdn.microsoft.com/en-us/library/141e06ef%28v=vs.71%29.aspx](Microsoft's guidelines) abbreviations should be in camel case (`consoleOrIisHostedAppRegex`)

File: src/RawRabbit/Common/NamingConventions.cs [hunk 2]
```diff
@@ -66,13 +67,34 @@ private string GetSubscriberQueueSuffix(Type messageType)
...
-				.Select(d => d.ToLower())
-				.Last(d => !_disallowedDirectoryNames.Contains(d))
-				.Split('.')
-				.Last();
+			return GetApplicationNameFromUrl(Directory.GetCurrentDirectory());
+		}
+
+		public string GetApplicationNameFromUrl(string url)
+		{
+			var consoleOrIISHostedAppRegex = new Regex(@".*\\(?<ApplicationName>.*[a-zA-Z][^\\]*)(\\[0-9.]*)?\\bin.*");
```
Comment (pardahlman) [inline src/RawRabbit/Common/NamingConventions.cs]: Could this be a private field only assigned once?

File: src/RawRabbit/Common/NamingConventions.cs [hunk 3]
```diff
@@ -66,13 +67,34 @@ private string GetSubscriberQueueSuffix(Type messageType)
...
+				match = windowsServiceRegex.Match(url);
+				if (match.Success)
+				{
+					applicationName = match.Groups["ApplicationName"].Value;
+				}
+			}
+
+			if (applicationName == string.Empty)
+			{
+				throw new IOException("Could not determine the application name");
```
Comment (pardahlman) [inline src/RawRabbit/Common/NamingConventions.cs]: Don't you think it is a bit harsh to throw here? We could return `string.Empty` or perhaps `unknown`?

Comment (Originalutter) [inline src/RawRabbit/Common/NamingConventions.cs]: You're right, let's just skip the string.Empty check

File: src/RawRabbit/Common/NamingConventions.cs (hunk 2 above)
Comment (Originalutter) [inline src/RawRabbit/Common/NamingConventions.cs]: Yeah, I guess that would be better!

File: src/RawRabbit/Common/NamingConventions.cs (hunk 1 above)
Comment (Originalutter) [inline src/RawRabbit/Common/NamingConventions.cs]: Nice resource, I always get confused when it comes to this :)

**Comment by pardahlman** [comment]
@Originalutter thanks! I've added a few comments and questions for the commit. Since the Application name should be part of the queue name, it makes sense to remove any dots (.) in the found name. I also realized that there is another delimitation: the queue name has a limited length. Let me think about this for a bit and get back to you.

**Comment by Originalutter** [comment]
I've fixed your comments. Still need to decide how to handle the limitation of the queue name length though.

File: src/RawRabbit/Common/NamingConventions.cs [hunk 4]
```diff
@@ -64,15 +66,40 @@ private string GetSubscriberQueueSuffix(Type messageType)
...
-			return Directory
-				.GetCurrentDirectory()
-				.Split('\\')
-				.Select(d => d.ToLower())
-				.Last(d => !_disallowedDirectoryNames.Contains(d))
-				.Split('.')
-				.Last();
+			return GetApplicationNameFromCommandLine(Environment.CommandLine);
+		}
+		public static string GetApplicationNameFromCommandLine(string commandLine)
```
Comment (pardahlman) [inline src/RawRabbit/Common/NamingConventions.cs]: Missing line break?

File: src/RawRabbit/Common/NamingConventions.cs [hunk 5]
```diff
@@ -64,15 +66,40 @@ private string GetSubscriberQueueSuffix(Type messageType)
...
+		private static string GetApplicationName()
 		{
-			return Directory
-				.GetCurrentDirectory()
-				.Split('\\')
-				.Select(d => d.ToLower())
-				.Last(d => !_disallowedDirectoryNames.Contains(d))
-				.Split('.')
-				.Last();
+			return GetApplicationNameFromCommandLine(Environment.CommandLine);
```
Comment (pardahlman) [inline src/RawRabbit/Common/NamingConventions.cs]: Why are we calling `GetApplicationNameFromCommandLine` here? IMHO `GetApplicationName` should contain the logic and be called with `Environment.CommandLine`

File: src/RawRabbit/Common/NamingConventions.cs (hunk 4 above)
Comment (Originalutter) [inline src/RawRabbit/Common/NamingConventions.cs]: Yes :)

File: src/RawRabbit/Common/NamingConventions.cs (hunk 5 above)
Comment (Originalutter) [inline src/RawRabbit/Common/NamingConventions.cs]: You're right! Earlier I did some preparation in `GetApplicationName` before calling `GetApplicationNameFromCommandLine` but as for now this is unnecessary. I will remove it :)

**Comment by pardahlman** [comment]
@Originalutter you've done it again. PR merged :+1:

**Pull request merged**